lib/msal-browser/src/interaction_client/PopupClient.ts

```diff
@@ -364,10 +267,187 @@ export class PopupClient extends StandardInteractionClient {
                 (e as AuthError).setCorrelationId(this.correlationId);
                 serverTelemetryManager.cacheFailedRequest(e);
             }
+
+            if (e instanceof ServerError && e.errorCode === BrowserConstants.INVALID_GRANT_ERROR) {
+                if (!this.requestRetried) {
+                    this.requestRetried = true;                
+                    if (!authClient || !authCodeRequest) {
+                        throw e;
+                    } else {
+                        return await invokeAsync(
+                            this.acquireTokenPopupAsyncHelper.bind(this),
+                            PerformanceEvents.PopupClientTokenHelper,
+                            this.logger,
+                            this.performanceClient,
+                            this.correlationId
+                        )(
+                            authClient,
+                            authCodeRequest,
+                            validRequest,
+                            request,
+                            popupName,
+                            popupWindowAttributes,
+                            popup
+                        )
+                    }
+                } else {
+                    // TODO: Log data point
+                    this.logger.trace("PopupClient.acquireTokenPopupAsync: requestRetried already true, throwing error");
+                    throw e;
+                }
+            }
+
             throw e;
         }
     }
 
+    /**
+     * Helper which obtains an access_token for your API via opening a popup window in the user's browser
+     * @param authClient 
+     * @param authCodeRequest 
+     * @param validRequest 
+     * @param request  
+     * @param popupName 
+     * @param popupWindowAttributes 
+     * @param popup 
+     * @returns A promise that is fulfilled when this function has completed, or rejected if an error was raised.
+     */
+    protected async acquireTokenPopupAsyncHelper(
```

Comment:
Can we please test observability logs too for the new helper function?

Reply:
Could you please elaborate what you mean by this?

Reply:
Observability is telemetry.  We should understand that a retry is happening from telemetry and note the original call result (invalid grant, possibly others in the future) as well as the result from the retry.  You can implement this by adding a datapoint that captures a RetryTriggerError (we can workshop the name) that has the failure in it (again invalid grant, someday maybe something else)  Presence of the error would indicate a retry was called, the final result is still logged in the existing telemetry so all existing telemetry flows are still correct.  That should cover what we need.   @konstantin-msft might have other suggestions for something more optimal though.

Reply:
Thanks for the input. I added a data point to log this.